### PR TITLE
Added support for ramp-pricing plans

### PR DESCRIPTION
**Pull request by Mozart2234**
- Adds `pricing_model` attribute to `Plan` Resource, which can be either 'fixed' (for fixed-pricing plans) or 'ramp' (for ramp-pricing plans)
- Adds `ramp_intervals` attribute to Plan Resource
- Adds `RampInterval` class, which can be used to initialize ramp_intervals for plans via the `ramp_intervals` attribute
- Adds unit tests

### CODE REVIEW

File: recurly/__init__.py [hunk 1]
```diff
@@ -1690,8 +1705,12 @@ class Plan(Resource):
         'auto_renew',
         'allow_any_item_on_subscriptions',
         'dunning_campaign_id',
+        'pricing_model',
+        'ramp_intervals',
     )
 
+    # _classes_for_nodename = {'ramp_intervals': RampInterval}
```
Comment (Patrick-Duvall) [inline recurly/__init__.py]: Is this comment mean to be here? I see the method elsewhere in the file but never commented out.

File: recurly/__init__.py [hunk 2]
```diff
@@ -1649,6 +1649,21 @@ def refund(self, **kwargs):
 Transaction._classes_for_nodename['transaction'] = Transaction
 
 
+class RampInterval(Resource):
+    """A plan ramp_interval
+
+    Uses this to represent a element of ramps_intervals
```
Comment (Patrick-Duvall) [inline recurly/__init__.py]: Does this need to be called `PlanRampInterval` to account for `SubRampInterval` later?

Looking at similar descriptions, maybe
```
"" A plan ramp, representing a price point and the billing_cycle to begin that price point""
```

File: recurly/resource.py [hunk 3]
```diff
@@ -445,11 +445,23 @@ def value_for_element(cls, elem):
...
 
         # Untyped complex elements should still be resource instances. Guess from the nodename.
-        if len(elem):  # has children
+        if len(elem) == 1:
             value_class = cls._subclass_for_nodename(elem.tag)
             log.debug("Converting %r tag into a %s", elem.tag, value_class.__name__)
             return value_class.from_element(elem)
 
+        # Untyped complex elements should still be resource instances. Guess from the nodename.
+        if len(elem) > 1:  # has childrens
```
Comment (Patrick-Duvall) [inline recurly/resource.py]: nit: children

File: recurly/resource.py [hunk 4]
```diff
@@ -445,11 +445,23 @@ def value_for_element(cls, elem):
                 return value_class.from_element(elem)
 
         # Untyped complex elements should still be resource instances. Guess from the nodename.
-        if len(elem):  # has children
+        if len(elem) == 1:
             value_class = cls._subclass_for_nodename(elem.tag)
             log.debug("Converting %r tag into a %s", elem.tag, value_class.__name__)
             return value_class.from_element(elem)
 
+        # Untyped complex elements should still be resource instances. Guess from the nodename.
```
Comment (Patrick-Duvall) [inline recurly/resource.py]: Maybe delete this duplicated comment and say something to the effect that we de-serialize ramp intervals when type="array" is not provided, i.e. with ramp_intervals.

File: recurly/__init__.py (hunk 1 above)
Comment (Mozart2234) [reply recurly/__init__.py]: you right

File: recurly/resource.py [hunk 5]
```diff
@@ -445,11 +445,23 @@ def value_for_element(cls, elem):
...
+        if len(elem) > 1:  # has childrens
+            first_tag = elem[0].tag
+            last_tag = elem[-1].tag
+
+            if(first_tag == last_tag):
+                return [cls._subclass_for_nodename(sub_elem.tag).from_element(sub_elem) for sub_elem in elem]
+            else:
+                value_class = cls._subclass_for_nodename(elem.tag)
+                log.debug("Converting %r tag into a %s", elem.tag, value_class.__name__)
+                return value_class.from_element(elem)
```
Comment (Smolations) [inline recurly/resource.py]: This seems like a lot of code to add to support this new feature. There should be other resources that have similar patterns to ramps, yea? Like, what about `add_ons`/`add_on`? That's a parent/many-child relationship isn't it? What is/are the reason(s) you needed to make this change?

Comment (Mozart2234) [reply recurly/resource.py]: as the definition of `subscription_add_ons` in `clients/python/tests/fixtures/subscription/show.xml`, it have the array definition, so the de-serialization happens in this line.

```python
# recurly/resource.py:433
    if attr_type == 'array':
       return [cls._subclass_for_nodename(sub_elem.tag).from_element(sub_elem) for sub_elem in elem]
```


```xml
<subscription_add_ons type="array">
    <subscription_add_on>
        <add_on_type>usage</add_on_type>
        <measured_unit href="https://api.recurly.com/v2/measured_units/123456"/>
        <usage href="https://api.recurly.com/v2/subscriptions/123456789012345678901234567890ab/add_ons/marketing_emails/usage"/>
        <add_on_code>marketing_emails</add_on_code>
        <unit_amount_in_cents type="integer">5</unit_amount_in_cents>
        <quantity type="integer">1</quantity>
        <usage_type>price</usage_type>
        <usage_percentage nil="nil"/>
    </subscription_add_on>
  </subscription_add_ons>
```
And in the ramp_intervals the `type` not exist.

```xml
  <ramp_intervals>
    <ramp_interval>
      <starting_billing_cycle type="integer">1</starting_billing_cycle>
      <unit_amount_in_cents>
        <USD type="integer">2000</USD>
      </unit_amount_in_cents>
    </ramp_interval>
    <ramp_interval>
      <starting_billing_cycle type="integer">2</starting_billing_cycle>
      <unit_amount_in_cents>
        <USD type="integer">3000</USD>
      </unit_amount_in_cents>
    </ramp_interval>
  </ramp_intervals>
```

Comment (Mozart2234) [reply recurly/resource.py]: in the file changes i left some example comments that I trying to de-serialize.

File: recurly/resource.py [hunk 6]
```diff
@@ -445,11 +445,34 @@ def value_for_element(cls, elem):
...
+            #   <item>...</item>
+            # </items>
+            if(first_tag == last_tag):
+                return [cls._subclass_for_nodename(sub_elem.tag).from_element(sub_elem) for sub_elem in elem]
+            # De-serialize one resource
+            # <resource>
+            #   <name>name</name>
+            #   <description>description</name>
+            #   <other_attribute>other text</other_description>
+            # </resource>
```
Comment (Patrick-Duvall) [inline recurly/resource.py:470]: This is quite verbose, but also super clear, i'm ok with it

**Review by Patrick-Duvall — APPROVED** [review]

**Pull request merged**